### PR TITLE
Add toolchain debug features to the depsearch tool

### PR DESCRIPTION
**Pull request by dmcilvaney**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [ ] Ready to merge

---

###### Summary 

Add tools to help debug toolchain build conflicts (ie the scheduler thinking it needs to rebuild toolchain packages, which it shouldn't)

###### Change Log  

- Add a few features to the depsearch tool:
 - `--tree` now outputs a simple tree view of the subgraph.
 - `--rpm-filter`  will limit the results to just those that aren't inside the filter list
 - `--rpm-filter-file` sets the filter for the above flag. Defaults to the x86_64 toolchain manifest if it can find it.
 - `--print-duplicates` sets the tree mode to print all nodes. By default it won't print a node if its already printed it once (warning, this takes a LONG time to run and can use up all your memory for even medium sized graphs, use with --max-depth)

- Added a section to the scheduler summary that lists all packages that were rebuild that conflict with the toolchain. This may become a breaking error in the future.
 
 ###### Does this affect the toolchain?  

**YES** (In that it may change the behaviour of toolchain packages as we are familiar with them since they will ALWAYS be the ones from the normal toolchain build)

### CODE REVIEW

File: toolkit/scripts/toolchain.mk [hunk 1]
```diff
@@ -19,6 +19,8 @@ populated_toolchain_chroot = $(toolchain_build_dir)/populated_toolchain
 toolchain_sources_dir = $(populated_toolchain_chroot)/usr/src/mariner/SOURCES
 populated_toolchain_rpms = $(populated_toolchain_chroot)/usr/src/mariner/RPMS
 toolchain_spec_list = $(toolchain_build_dir)/toolchain_specs.txt
+toolchain_actual_conents = $(toolchain_build_dir)/actual_archive_contents.txt
```
Comment (Camelron) [inline toolkit/scripts/toolchain.mk:22]: Should this be "toolchain_actual_contents"?

File: toolkit/scripts/toolchain.mk [hunk 2]
```diff
@@ -19,6 +19,8 @@ populated_toolchain_chroot = $(toolchain_build_dir)/populated_toolchain
...
+toolchain_expected_conents = $(toolchain_build_dir)/expected_archive_contents.txt
```
Comment (Camelron) [inline toolkit/scripts/toolchain.mk:23]: likewise, toolchain_expected_contents

File: toolkit/scripts/toolchain.mk (hunk 1 above)
Comment (dmcilvaney) [reply toolkit/scripts/toolchain.mk]: Fixed in the makefile fixup PR

File: toolkit/scripts/toolchain.mk (hunk 2 above)
Comment (dmcilvaney) [reply toolkit/scripts/toolchain.mk]: Fixed in the makefile fixup PR

**Comment by dmcilvaney** [comment]
Done WIP work, now just the depsearch tool changes.

File: toolkit/tools/depsearch/depsearch.go [hunk 3]
```diff
@@ -166,16 +225,235 @@ func buildDependsOnGraph(graphIn *pkggraph.PkgGraph, nodeList []*pkggraph.PkgNod
 	return
 }
 
-func printSpecs(graph *pkggraph.PkgGraph) {
-	specs := make(map[string]bool)
-	for _, n := range graph.AllNodes() {
+const (
+	colourReset = "\033[0m"
```
Comment (hbeberman) [inline toolkit/tools/depsearch/depsearch.go:229]: 🍁

**Review by hbeberman — APPROVED** [review]

**Pull request merged**